File: backend/schemas/leaves.py
from datetime import date, datetime

from pydantic import BaseModel, Field

from models.leaves import LeaveStatus, LeaveType


class LeaveRequestCreate(BaseModel):
    leave_on: date
    leave_type: LeaveType
    user_compensation_id: int | None = None


class LeaveDecisionBody(BaseModel):
    motivo: str | None = Field(default=None, max_length=4000)


class LeaveRejectBody(BaseModel):
    motivo: str = Field(..., min_length=1, max_length=4000)


class CalendarLeaveEntry(BaseModel):
    id: int
    leave_on: date
    user_id: int
    patente: str
    nome_guerra: str
    display_order: int
    leave_type: LeaveType
    status: LeaveStatus
    operational_rank: int

    model_config = {"from_attributes": False}


class CalendarDay(BaseModel):
    date: date
    entries: list[CalendarLeaveEntry]
    active_count: int
    is_critical: bool


class LeaveCalendarSummary(BaseModel):
    my_pending_count: int
    command_pending_leaves: int | None = None
    command_pending_compensations: int | None = None
    critical_days: list[date] | None = None


class YearMonth(BaseModel):
    year: int
    month: int


class LeaveBookingPolicy(BaseModel):
    reference_date: date
    allowed_year_months: list[YearMonth]
    operational_hint: str


class LeaveCalendarResponse(BaseModel):
    year: int
    month: int
    days: list[CalendarDay]
    summary: LeaveCalendarSummary
    booking_policy: LeaveBookingPolicy


class LeaveRequestPublic(BaseModel):
    id: int
    user_id: int
    leave_on: date
    leave_type: LeaveType
    user_compensation_id: int | None
    status: LeaveStatus
    review_reason: str | None
    decision_motivo: str | None
    decided_by_id: int | None
    decided_at: datetime | None
    created_at: datetime
    patente: str | None = None
    nome_guerra: str | None = None
    display_order: int | None = None

    model_config = {"from_attributes": False}

File: frontend/src/types/leaves.ts
export type LeaveType = "MONTHLY" | "COMPENSATION" | "DS";

export type LeaveStatus = "PENDING" | "REVIEW" | "APPROVED" | "REJECTED" | "CANCELLED";

export type {
  CompensationType,
  CompensationEventStatus,
  CompensationEventPublic,
} from "@/types/compensations";

export type UserCompensationStatus = "AVAILABLE" | "USED";

export interface CalendarLeaveEntry {
  id: number;
  leave_on: string;
  user_id: number;
  patente: string;
  nome_guerra: string;
  display_order: number;
  leave_type: LeaveType;
  status: LeaveStatus;
  operational_rank: number;
}

export interface CalendarDay {
  date: string;
  entries: CalendarLeaveEntry[];
  active_count: number;
  is_critical: boolean;
}

export interface LeaveCalendarSummary {
  my_pending_count: number;
  command_pending_leaves?: number | null;
  command_pending_compensations?: number | null;
  critical_days?: string[] | null;
}

export interface YearMonth {
  year: number;
  month: number;
}

export interface LeaveBookingPolicy {
  reference_date: string;
  allowed_year_months: YearMonth[];
  operational_hint: string;
}

export interface LeaveCalendarResponse {
  year: number;
  month: number;
  days: CalendarDay[];
  summary: LeaveCalendarSummary;
  booking_policy: LeaveBookingPolicy;
}

export interface LeaveRequestPublic {
  id: number;
  user_id: number;
  leave_on: string;
  leave_type: LeaveType;
  user_compensation_id: number | null;
  status: LeaveStatus;
  review_reason: string | null;
  decision_motivo: string | null;
  decided_by_id: number | null;
  decided_at: string | null;
  created_at: string;
  patente?: string | null;
  nome_guerra?: string | null;
  display_order?: number | null;
}

export interface UserCompensationPublic {
  id: number;
  user_id: number;
  compensation_event_id: number;
  status: UserCompensationStatus;
  created_at: string;
  display_label?: string;
}

/** Crédito disponível retornado por GET /compensations/available */
export interface UserCompensationAvailable {
  id: number;
  type: CompensationType;
  label: string;
  event_date: string;
  description: string;
}
